### PR TITLE
stage2: add support for optionals in the LLVM backend

### PR DESCRIPTION
**Pull request by FireFox317**
We can now codegen optionals! This includes the following instructions:
- isnull
- isnonnull
- unwrap_optional
- brvoid

Also includes a test.

The test currently does not pass because of #7730.

EDIT: Also sneak in an extra test for loops.

### CODE REVIEW

File: src/codegen/llvm.zig [hunk 1]
```diff
@@ -568,6 +580,40 @@ pub const LLVMIRModule = struct {
...
+
+        const index_type = self.context.intType(32);
+
+        var indices: [2]*const llvm.Value = .{
+            index_type.constNull(),
+            index_type.constInt(1, false),
+        };
+
+        return self.builder.buildLoad(self.builder.buildInBoundsGEP(operand, &indices, 2, ""), "");
+    }
```
Comment (FireFox317) [inline src/codegen/llvm.zig:602]: @Vexu why do these 2 instructions receive different types of operands? (optional vs pointer to optional)

Comment (Vexu) [reply src/codegen/llvm.zig:602]: I just copied what stage1 does. There are some TODOs regarding this in `astgen.zig`.

**Comment by g-w1** [comment]
Why does the CI pass? Is it not running the LLVM tests?

**Comment by FireFox317** [comment]
@g-w1 yep correct, the ci currently does not run the llvm tests (the `-Denable-llvm` flag is not passed)

**Comment by andrewrk** [comment]
Can you rebase this now that 7e56028bc7c00b884c92e2948728cbc47e5a8a09 has landed please?

**Review by andrewrk — CHANGES_REQUESTED** [review]
Can you rebase this now that 7e56028bc7c00b884c92e2948728cbc47e5a8a09 has landed please?

**Comment by FireFox317** [comment]
@andrewrk PR has been rebased, however the current test does not pass semantic analysis.
```zig
var opt_val: ?i32 = 10;
const val: i32 = opt_val.?;
```
ZIR:
```
fn_body main { // unanalyzed
  %0 = dbg_stmt()
  %1 = const(TypedValue{ .ty = type, .val = type})
  %2 = const(TypedValue{ .ty = type, .val = type})
  %3 = const(TypedValue{ .ty = type, .val = i32})
  %4 = as(%2, %3)
  %5 = optional_type(%4)
  %6 = as(%1, %5)
  %7 = alloc_mut(%6)
  %8 = const(TypedValue{ .ty = comptime_int, .val = 10})
  %9 = coerce_to_ptr_elem(%7, %8)
  %10 = store(%7, %9)
  %11 = dbg_stmt()
  %12 = const(TypedValue{ .ty = type, .val = type})
  %13 = const(TypedValue{ .ty = type, .val = i32})
  %14 = as(%12, %13)
  %15 = deref(%7)
  %16 = as(%14, %15) <------------------------------------
  %17 = optional_payload_safe(%16)
  %18 = dbg_stmt()
  %19 = ret_type()
  %20 = const(TypedValue{ .ty = comptime_int, .val = 0})
  %21 = as(%19, %20)
  %22 = return(%21)
} // fn_body main
```
The error happens in instruction `%16`. The `as` is emitted as part of the `rlWrap`, because of line 921 in astgen.zig.

Then the question is whether `Module.coerce` should be able to go from a `?T -> T`, or if the `as` (or `coerce_to_ptr_elem` for `var`) instruction should not be emitted at all.

**Comment by FireFox317** [comment]
PR has been rebased and is ready for review. Please carefully review `astgen.zig`, because I'm not sure if this is the correct solution.

**Comment by FireFox317** [comment]
@andrewrk PR has been rebased after ast changes. Again please carefully review the changes in `astgen.zig`, because i'm not sure if they are correct (it does fix compilation and all the tests are passing!)

I also include a commit that fixes some fmt changes that were missed in cd7c870bd81391dd97c5c75eb3910382ba7280a1

**Pull request merged**